### ISSUE TITLE
MSE incorrect when using fit_generator and max_queue_size != 1

### ISSUE THREAD

**Issue by alekseynp** [keras-team/keras#8647]
I am getting unexpected behaviour when using mean_squared_error loss.

My original issue was that fit_generator was reporting an incorrect validation loss. I was unable to reproduce that bug in a gist that I can make public, but I was able to demonstrate an incorrect training loss.

Bafflingly, if I change max_queue_size to 1 instead of the default 10 in fit_generator, the issue is resolved.

https://gist.github.com/alekseynp/1bb6cbe4ee7fb6e8ea46d68c675b934d

**Comment by fchollet** [comment]
Not sure what you really mean (you don't provide much info), but note that the docs say that using `fit_generator` with multiple workers and a Python generator that isn't a `Sequence` instance, can result in duplicated data. Maybe that's what you are observing.

To avoid such data duplication issues, you can either use a single worker, or implement a `Sequence` instead of a generator to produce the data to train your model.

**Comment by alekseynp** [comment]
Sorry for not providing more info. I hoped that the gist would be demonstrative. I thought this was an MSE-related error, but that seems to not be the case now. I'll admit it seems kind of crazy that it would be MSE-specific.

I tried implementing a `Sequence` and was successful. This led me to review my code that has the issue, so here's some more analysis.

`fit_generator` by default uses `max_queue_size=10`, `workers=1`, and `use_multiprocessing=False` which leads me to believe that I do not need to take special steps to make my code "safe" for multiprocessing. However, this is not the case.

When I use the following generator and the default settings for `fit_generator`, I have problems which I see manifest in incorrectly reported training_loss and validation_loss.
```
def my_generator(batch_size):
    inputs = np.zeros((batch_size,10), dtype=np.float32)
    outputs = np.zeros((batch_size,1), dtype=np.float32)

    while True:
        for i in range(batch_size):
            inputs[i,:] = np.random.random(10)
            outputs[i,0] = np.sum(np.arange(0,10) * inputs[i,:])

        yield inputs, outputs
```
If I change `max_queue_size` to `1`, my problems are resolved.


When I move the array creation inside the `while True` loop, my issue is also resolved.
```
def my_generator(batch_size):
    while True:
        inputs = np.zeros((batch_size,10), dtype=np.float32)
        outputs = np.zeros((batch_size,1), dtype=np.float32)

        for i in range(batch_size):
            inputs[i,:] = np.random.random(10)
            outputs[i,0] = np.sum(np.arange(0,10) * inputs[i,:])

        yield inputs, outputs
```

So obviously for resolving my own problem, the key is to not re-use the same array in my generator. And I would hope that other people with issues will find this and find the same resolution. From a Keras perspective, this seems like a gotcha for users falsely thinking that they are insulated from problems related to multi-processing.
